js/count_characters.js

```diff
@@ -38,4 +70,71 @@ function addCount(obj) {
 function getParagraphs() {
     var paragraphs = document.getElementsByClassName("exchange--speaker");
     return paragraphs;
-};
+}
+
+
+
+function update_subtitle(){
+    var current_time = time_bar.getElementsByClassName("timecode text--noselect")[0];
```

Comment:
Here's an example that might help explain..
![milliseconds_time](https://user-images.githubusercontent.com/33587724/72844936-3afb3200-3ca6-11ea-9a07-3b18be5a1f0b.png)

Reply:
I agree.
I get the time now form the width of the video progress bar (width, style property) which is % and float..
the paragraph start time, I now get it as you mention.
However, the the subtitle is still out of sync (but very negligible) because the event that calls the caption update, is fired every second (since it captures the time of the left).
I don't think this is bad, if the event is fired more often (10 times a second), the add-on might become heavy on slow computers! not sure about that, but there is a point here :sweat_smile:

Reply:
Seems great, I wouldn't think it would be an issue, even if you made it refresh every 50 milliseconds (No need for the observer).
What you could do is save a global `previous_index` indicating the latest used index so you don't update the caption if the index did not change.. which might reduce the computations.
Also, I'm not sure if writing the helper functions inside a function that is frequently called will re-define them, thus taking more time each call.. I prefer if you write them outside the function as independent function.

Reply:
I changed the even now from observer to ```ontimeupdate```.
The out of sync (which is very negligible) is still there, even if I try a timed event (every 100milis), it does not improve the result. But the offset is very small, and I think it won't user experience

Reply:
I tried using this method: https://stackoverflow.com/questions/2170923/whats-the-easiest-way-to-call-a-function-every-5-seconds-in-jquery
By wrapping `update_caption` with this native js timer with 50 ms, looks much smoother for me..
Could you try it and let me know how it looks for you?

Reply:
ok, I thought about something similar too... I'll check it now..

Reply:
it seems the problem is the calculated current time (video progress bar)...

Reply:
I changed how current time is obtained. It solved an issue I was facing when a paragraph is very short.
I tried the timer based event (10 ms), it doesn't make a difference. According to [this](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/timeupdate_event), the event should be fired at max 66Hz (15 ms).

Reply:
Looks nice!
I just noticed something that could be an easy but also nice update..
when you type any characters even newline, time is not changed thus the event is not triggered.
Can we trigger the event with every mouse click on keyboard button pressed?

Reply:
I agree, it will look better that way..
for the end time issue, I just tried it, the thing is, Sonix seems to assign the start time of the next paragraph same as the end time of the previous one.. so there is no need to do that..